### PR TITLE
Add `set_meta_from_data()` feature

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

This PR implements the feature requested by @znicholls (see #238 and #237) to derive meta indicators directly from the `data` dataframe, with a `method` arg passed for aggregation (e.g., return the year where the temperature maximum is reached).

# Discussion item

It might be more user-friendly to allow strings as `method` arg in addition to functions, i.e., `method='max'` rather than `method=`np.max`. Any thoughts?

Closes 238

### CODE REVIEW

**Comment by danielhuppmann** [comment]
@znicholls, can I please get a hug for writing a test that also ensures it works with `datetime`? ;-)

**Comment by danielhuppmann** [comment]
unit tests are failing because of issues with the region-plotting, not the new feature and related tests

for the review, it makes most sense to only look at commit https://github.com/IAMconsortium/pyam/pull/241/commits/f5151892cb760c79eb6b7e6848a225600546ec7c, the commits before that are just clean-up and moving all tests related to setting meta indicators to an own file

**Comment by znicholls** [comment]
I send you all the hugs! I’ll do a proper review Monday, for now my only thought is whether applying `method` with `pd.groupby.apply(method)` rather than `method(pd.groupby)` would provide more flexibility.

File: pyam/core.py [hunk 1]
```diff
@@ -458,6 +458,27 @@ def set_meta(self, meta, name=None, index=None):
...
+            unique values (e.g., `numpy.max()`)
+        column: str, optional
+            the column from `data` to be used for the meta indicator
+        kwargs: passed to :meth:`IamDataFrame.filter()` for downselected `data`
+        """
+        _data = self.filter(**kwargs).data
+        if method is None:
+            meta = _data.set_index(META_IDX)[column]
+        else:
+            meta = method(_data.groupby(META_IDX))[column]
```
Comment (znicholls) [inline pyam/core.py]:
```suggestion
            meta = _data.groupby(META_IDX).apply(method)[column]
```

alternate way which might provide more flexibility

Comment (gidden) [reply pyam/core.py]: Perhaps even `meta = _data.groupby(META_IDX)[column].apply(method)`?

**Comment by gidden** [comment]
hey folks, trying out now a fix for CI (hopefully permanent..). Regarding how to pass the `method` arg, I'm a bit torn here, as I can see arguments in both cases (passing a function or string). One thing that is nice about using a string is that it plays well with more on-disk usage (e.g., writing a config file, being able to pickle/serialize objects, etc.).  However, it seems like the work is done and is quite straightforward for using `apply`.

Perhaps we can leave the question open and potentially support both in the future. In that case, I would change the kwarg from `method` to `apply`. This provides clarity to the user (i.e., `np.max` will be used by `df.apply`). It also leaves us the option of in the future providing a more generic `method` arg, which would simply take a dataframe as argument.

**Comment by gidden** [comment]
Hi folks - CI has been updated on the master branch. Please rebase this one on top!

**Comment by gidden** [comment]
Restarted the build, should bring in updates from master

File: pyam/core.py (hunk 1 above)
Comment (danielhuppmann) [reply pyam/core.py]: done

**Comment by danielhuppmann** [comment]
rebased and refactored to `data.apply(method)` as suggested by @znicholls (though this does still not allow to use `np.mean()` as method because this doesn't work on `pandas.groupby`)

**Comment by znicholls** [comment]
> though this does still not allow to use `np.mean()` as method because this doesn't work on `pandas.groupby`

this test passes ?

```python
def test_set_meta_from_data_mean(meta_df):
    meta_df.set_meta_from_data('pe_mean', variable='Primary Energy', method=np.mean)
    exp = pd.Series(data=[3.5, 4.5], index=EXP_IDX, name='pe_mean')
    pd.testing.assert_series_equal(meta_df['pe_mean'], exp)
```

**Comment by danielhuppmann** [comment]
> this test passes ?

Seems to pass now - no idea why it failed yesterday when I tried it...

**Comment by znicholls** [comment]
@gidden lgtm so over to you for merging

**Comment by gidden** [comment]
LGTM, should we add this feature in a tutorial somewhere?

**Comment by danielhuppmann** [comment]
added a section in the 'first-steps' tutorial per suggestion by @gidden

**Comment by danielhuppmann** [comment]
reopening to get pandas 0.25.0 hotfix after merging #246

**Comment by gidden** [comment]
Thanks @danielhuppmann !

**Pull request merged**